### ISSUE TITLE
Cannot reproduce the probe result (Figure 2: (A) in the paper)

### ISSUE THREAD

**Issue by jiahaozhenbang** [likenneth/honest_llama#43]
Hi,

thanks for putting this helpful repo together.

I've been trying to reproduce the probe result (Figure 2: (A) in the paper) and run the code exactly as described in the [workflow](https://github.com/likenneth/honest_llama?tab=readme-ov-file#workflow):

```
#  Get activations
python get_activations.py --model_name llama_7B --dataset_name tqa_gen_end_q
python get_activations.py --model_name llama_7B --dataset_name tqa_mc2

# Probe activations
CUDA_VISIBLE_DEVICES=1 python validate_2fold.py --model_name llama_7B --num_heads 48 --alpha 15
 --device 0 --num_fold 2 --use_center_of_mass --instruction_prompt default --judge_name <your GPT-judge name> --info_name <your GPT-info name>
```
Here, I run the code until the function [`get_top_heads`](https://github.com/likenneth/honest_llama/blob/master/validation/validate_2fold.py#L143), save the probing accuracy on each head (a total of 2\*32\*32), average the accuracy for two-fold and plot the heatmap as follows.
![output](https://github.com/user-attachments/assets/9efd8d0a-4f57-493d-8277-2fc428b7633a)
The top accuracy, for example, is achieved by the 12th head in the 11th layer, which has a validation accuracy of 79.2%.

However, this differs from Figure 2: (A) in the paper. Do you have any ideas or can you provide more details?

Thanks much in advance.

**Comment by likenneth** [comment]
validate_two_fold.py is for validating the ITI performances on a set of heads, say the top 48. It's not obvious to me why it could give you a matrix of 2*32*32.


The way I produced Figure 2A is by writing another simple script to run logistic regression on each head's activation extracted by get_activations.py, yielding a 32*32 matrix of validation accuracy.

**Comment by jiahaozhenbang** [comment]
Thank you for your reply.

I am sorry that my unclear statement caused you a misunderstanding. I would like to clarify here. I made some changes to validate_2fold.py, keeping only the part that calculates the probe accuracy, and did not conduct further evaluation. 

For the sake of clarity, I post the key code here as follows：

```
# load activations 
head_wise_activations = np.load(f"../features/{args.model_name}_{args.dataset_name}_head_wise.npy")
labels = np.load(f"../features/{args.model_name}_{args.dataset_name}_labels.npy")
head_wise_activations = rearrange(head_wise_activations, 'b l (h d) -> b l h d', h = num_heads)


separated_head_wise_activations, separated_labels, idxs_to_split_at = get_separated_activations(labels, head_wise_activations)
# run k-fold cross validation
results = []
for i in range(args.num_fold):

    train_idxs = np.concatenate([fold_idxs[j] for j in range(args.num_fold) if j != i])
    test_idxs = fold_idxs[i]

    print(f"Running fold {i}")

    # pick a val set using numpy
    train_set_idxs = np.random.choice(train_idxs, size=int(len(train_idxs)*(1-args.val_ratio)), replace=False)
    val_set_idxs = np.array([x for x in train_idxs if x not in train_set_idxs])

    # save train and test splits
    df.iloc[train_set_idxs].to_csv(f"splits/fold_{i}_train_seed_{args.seed}.csv", index=False)
    df.iloc[val_set_idxs].to_csv(f"splits/fold_{i}_val_seed_{args.seed}.csv", index=False)
    df.iloc[test_idxs].to_csv(f"splits/fold_{i}_test_seed_{args.seed}.csv", index=False)

    probes, all_head_accs_np = train_probes(args.seed, train_set_idxs, val_set_idxs, separated_head_wise_activations, separated_labels, num_layers=num_layers, num_heads=num_heads)
    all_head_accs_np = all_head_accs_np.reshape(num_layers, num_heads)

    print(f"FOLD {i}")
    print(all_head_accs_np)

    all_head_accs_np = all_head_accs_np.astype(float)
    results.append(all_head_accs_np)

results = np.array(results)
final = results.mean(axis=0)
```

It sounds like the same implementation as yours, but the visualization results are somewhat different. I wonder if you have any insights or suggestions.

Sincerely looking forward to your reply!

**Comment by likenneth** [comment]
Maybe the problem is that the train/val split is 50% here while more training data is used for making Figure 2A. 50% is used here solely for a rigorous 2-fold evaluation. To figure out which heads have a more learnt representation of truthfulness, a 4:1 split could give a more salient heatmap in Figure 2A. 

From the paper: "We then randomly split each dataset into training and validation sets by 4 : 1".

**Comment by jiahaozhenbang** [comment]
Thanks for your reply! I will have a try.

**Comment by lixali** [comment]
In the above, @likenneth mentions **"The way I produced Figure 2A is by writing another simple script to run logistic regression on each head's activation extracted by get_activations.py, yielding a 32*32 matrix of validation accuracy."** 

I wonder can @likenneth  or @jujipotle  provide such script so that we can reproduce the Figure 2(A) in the paper?
